### DISCUSSION TITLE
Access to Proxy Headers / Connect Response

### DISCUSSION THREAD

**Discussion by chribro88** [encode/httpx#2862] [General]
Just checking if there's any way to access the proxy headers from the connect response ?

or is it required for changes to be made to httpcore e.g. [here](https://github.com/chribro88/httpcore/blob/feature/access_connect_response/httpcore/_async/http_proxy.py#L265) and [here](https://github.com/chribro88/httpcore/blob/feature/access_connect_response/httpcore/_async/http_proxy.py#L294) which would then expose the connect response to the client via 
```
proxy_headers = client._transport._pool.connections[0]._connect_response.headers
```

**Comment by lovelydinosaur** [comment]
> Just checking if there's any way to access the proxy headers from the connect response ?

There's no documented way to access the proxy headers from the connect response.

> which would then expose the connect response to the client via

We wouldn't expose the connect response via private API as in the example you've given, no.

If we *did* want to make this available, we'd probably want to use the response extensions API.
Perhaps something like...

```python
response.extensions['proxy_headers']
```

**Reply by chribro88** [reply]
Ok - that's good to know. Appreciate the quick reply.

For my current use case that way seems to be working but I'll definitely have a look at the extensions / request API and see if there's a better way to implement.

Would there be any examples of the request API you would recommend?

**Reply by lovelydinosaur** [reply]
> For my current use case that way seems to be working

Ok. You're digging into implementation details there which could change, so make sure to pin the package versions exactly.

> I'll definitely have a look at the extensions / request API and see if there's a better way to implement.

Sure thing. Best place to start isn't the implementation, but with an API design discussion.
If you're motivated to get this landed we can definitely make the time to guide you through it.

**Reply by cono** [reply]
also interested in this functionality!

**Reply by karpetrosyan** [reply]
You can take a look at the extensions that httpx supports. 

Maybe something like this could be useful:

```python
import httpx


def trace_func(event, data):
    if "receive_response_headers.started" in event:
        request = data["request"]
        print(request)
    elif "receive_response_headers.complete" in event:
        version, status, reason, headers = data["return_value"]

        print(version, status, reason)
        print(headers)


client = httpx.Client(proxies="https://yourproxy")

response = client.request(
    "GET", "https://www.encode.io", extensions={"trace": trace_func}
)
print(response)
```
